docs/DeveloperGuide.adoc

```diff
@@ -227,20 +250,159 @@ image::CommitActivityDiagram.png[]
 
-_{Explain here how the data encryption feature will be implemented}_
+====
+1. Create a Java file inside a temporary directory in the user's file system and writing the user's program to it.
 
-// end::dataencryption[]
+2. Compile that Java file into an executable Java Class file. Catch and store compile errors if any.
+
+3. Execute the main Class file and provide the inputs of the specified test cases. Collect and store any output from the program.
+
+4. Encapsulate the output, errors and results in a model class
+====
+
+This is done by `TestExecutor` with the help of three interfaces that are made available through dependency injection --
+`CompilerEnvironment`, `Compiler`, `ProgramExecutor`.
+
+Each interface has its specialized role in the evaluation process.
+`TestExecutor` acts as an orchestrator to tie all their functionalities together.
+
+====
+*CompilerEnvironment* - in charge of creating a temporary folder in the user's file system to create and write Java files.
+This temporary folder is deleted after `closed()`. Utilizes the model class `JavaFile`.
+
+image::CompilerEnvironment.png[align=center]
+
+*Compiler* - in charge of compiling the Java files into Class files at a given file path. Utilizes the model class `ClassFile`.
+
+image::Compiler.png[align=center]
+
+*ProgramExecutor* - in charge of executing the compiled Class files. Utilizes the model class `ProgramOutput`.
+
+image::ProgramExecutor.png[align=center]
+
+*Models:*
+
+* *JavaFile* - contains the canonical name and class path of a Java file with various convenience methods. Note that the file must actually exist or
+an `FileNotFoundException` is thrown during instantiation.
+* *ClassFile* - contains the canonical name and class path of a Java file with various convenience methods. Note that the file must actually exist or
+an `FileNotFoundException` is thrown during instantiation.
+* *ProgramInput* - contains the String input to be fed into the user's program.
+* *ProgramOutput* - contains the String output produced by the user's program. It also contains convenience methods for producing different outputs.
+====
+
+====
+*Sequence diagram:*
+
+1. `TestExecutor` calls `clearEnvironment()` of `CompilerEnvironment` to remove any redundant files from previous program
+evaluations.
+
+2. `TestExecutor` calls `createJavaFile()` of `CompilerEnvironment` to create the Java file with the class name and source code specified
+in the `UserProgram` argument.
+
+3. `TestExecutor` calls `compileJavaFile()` of `Compiler` to compile the newly created Java file.
+
+4. For each test case passed in the argument, `TestExecutor` calls `executeProgram` of `StandardProgramExecutor` with the input from the
+test case.
+
+5. `TestExecutor` maps each `ProgramOutput` received into a `TestCaseResult` model instance.
+
+6. `TestExecutor` packs all the data collected from the execution of the test cases into a single `TestResult` instance.
+
+image::TestExecutorSequenceDiagram.png[align:center]
```

Comment:
There should be colons before all the class names.
Return arrows should be dotted arrows.
Activation bar should stop after return.
Loop [test cases] can be more clear; like loop [while there are more test cases].
Why are there class names are the bottom of the diagram?